### PR TITLE
benchmark across git history

### PR DESCRIPTION
**Pull request by alexlamsl**
Looks like there are some commits that have infinite loop bugs, so will have to set up a watchdog timer to work around those buggy ones.

So far I can go back to [200 commits](https://github.com/kangax/html-minifier/files/129307/backtest.csv.txt).

### CODE REVIEW

File: backtest.js [hunk 1]
```diff
@@ -0,0 +1,179 @@
+#!/usr/bin/env node
+
+'use strict';
+
+var child_process = require('child_process'),
+    fs = require('fs'),
+    path = require('path');
+
+var urls = require('./benchmarks');
```
Comment (kangax) [inline backtest.js:11]: I was just writing about reusing urls between benchmark and backtest :)

Comment (alexlamsl) [inline backtest.js:11]: Some saying about great minds huh? :wink:

**Comment by alexlamsl** [comment]
Looking back at the commit history, it seems like using `src/htmlminifier.js` directly would be a better solution here...

**Comment by alexlamsl** [comment]
[Full git history run](https://github.com/kangax/html-minifier/files/129554/backtest.zip)

The earlier commits are not completing within one minute over here (i7-4960X @4.5GHz), but otherwise we/I don't seem to introduce too much regression thus far :sweat_smile:

**Comment by kangax** [comment]
Interesting. Not exactly what I expected (gradual decrease) but great to have nevertheless. Do you know why the spikes and higher current all-time filesize on some of them?

**Comment by alexlamsl** [comment]
That's what I'm investigating atm - I think some of them have genuine reasons, but others are just due to the various I/O (race) bugs in Node.js (libuv?), which have been illusive and platform-dependent (read: Windows :cry:)

**Comment by alexlamsl** [comment]
Just pasting a few bugs I've encountered so far:

e2e4ff5, a8d5bc3, ..., 2ddfee6

```
[amazon] TypeError: Object doesn't support property or method 'replace'
   at collapseWhitespaceSmart (src/htmlminifier.js:55:7)
   at chars (src/htmlminifier.js:554:13)
   at global.HTMLParser (src/htmlparser.js:221:13)
   at minify (src/htmlminifier.js:456:5)
   at Anonymous function (backtest.js:43:11)
   at tryToString (fs.js:414:3)
   at readFileAfterClose (fs.js:401:5)
```

d1d4635, 1127285, ..., 49b9e8d

```
[newyorktimes] Parse Error:

</body>
</html>
```

a397362, 34a7230, ..., 4615d6b

```
[*] TypeError: Unable to get property 'replace' of undefined or null reference
   at collapseWhitespaceSmart (src/htmlminifier.js:55:7)
   at chars (src/htmlminifier.js:553:13)
   at global.HTMLParser (src/htmlparser.js:155:13)
   at minify (src/htmlminifier.js:455:5)
   at Anonymous function (backtest.js:43:11)
   at tryToString (fs.js:414:3)
   at readFileAfterClose (fs.js:401:5)
```

d9b302d, f3fc9d1, ..., 110b540

```
[*] ReferenceError: 'HTMLParser' is undefined
   at minify (src/htmlminifier.js:262:5)
   at Anonymous function (backtest.js:43:11)
   at tryToString (fs.js:414:3)
   at readFileAfterClose (fs.js:401:5)
```

32b02f0, 10291f8, ..., 4705f91

```
[*] TypeError: Object expected
   at Anonymous function (backtest.js:43:11)
   at tryToString (fs.js:414:3)
   at readFileAfterClose (fs.js:401:5)
```

**Comment by alexlamsl** [comment]
These are the ones which are currently suffering a minfied size regression:
- es6-table
- html-minifier
- nbc
- stackoverflow

I suspect they are primarily due to the now safer behaviour of `removeOptionalTags`, where the more aggressive side of the latest changes is not enough to offset the growth.

**Comment by alexlamsl** [comment]
[Full history run](https://github.com/kangax/html-minifier/files/129691/backtest.zip) with higher success ratio this time.

Many of the graphs now have longer history (since they fail less often), and I think we are starting to see more significant improvements over time.

Obviously there is this caveat that we aren't measuring correctness by any means, which means if we rely of this metric alone `function minify() {return ''}` would be the clear winner :laughing:

**Comment by alexlamsl** [comment]
[backtest.zip](https://github.com/kangax/html-minifier/files/129789/backtest.zip)

**Comment by alexlamsl** [comment]
```
310 - [newyorktimes] Parse Error:

158 - [nbc] TypeError: Object doesn't support property or method 'replace'

 13 - [html-minifier] TypeError: Unable to get property 'replace' of undefined or null reference

 14 - [html-minifier] Error: unexpected result: undefined

 27 - [nbc] Parse Error: <div id="node_dl_slide_nbc_dynamic_slide_group_dl_wrap"class = "dynamic-lead-slide-content">
```

473 counts of `task timed out`, though most of them have some benchmarks that came through

**Comment by alexlamsl** [comment]
[backtest.zip](https://github.com/kangax/html-minifier/files/129842/backtest.zip)

@kangax I think this code is good to go now, PTAL :wink:

**Comment by alexlamsl** [comment]
@kangax any luck with reviewing? :sweat_smile:

**Pull request merged**